src/jitsu.ml

```diff
@@ -16,314 +16,283 @@
 
-let create_vm t vm =
-  try_libvirt "Unable to create VM" (fun () -> Libvirt.Domain.create (lookup_uuid t vm.vm_uuid))
+  let get_vm_state t vm =
+    or_vm_backend_error "Unable to get VM state from backend" (Vm_backend.get_state t.vm_backend) vm
 
-let stop_vm t vm =
-  match get_vm_state t vm with
-  | Libvirt.Domain.InfoRunning ->
-    begin match vm.how_to_stop with
-      | VmStopShutdown -> t.log (Printf.sprintf "VM shutdown: %s\n" vm.vm_name); shutdown_vm t vm
-      | VmStopSuspend  -> t.log (Printf.sprintf "VM suspend: %s\n" vm.vm_name) ; suspend_vm t vm
-      | VmStopDestroy  -> t.log (Printf.sprintf "VM destroy: %s\n" vm.vm_name) ; destroy_vm t vm
-    end
-  | _ -> ()
+  let stop_vm t vm =
+    get_vm_name t vm >>= fun vm_name ->
+    get_vm_state t vm >>= fun vm_state ->
+    match vm_state with
+    | Vm_state.Running ->
+      Irmin_backend.get_stop_mode t.storage ~vm_name >>= fun stop_mode ->
+      begin match stop_mode with
+        | Vm_stop_mode.Unknown -> t.log (Printf.sprintf "Unable to stop VM %s. Unknown stop mode requested\n" vm_name); 
+          Lwt.return_unit
+        | Vm_stop_mode.Shutdown -> t.log (Printf.sprintf "VM shutdown: %s\n" vm_name);
+          or_vm_backend_error "Unable to shutdown VM" (Vm_backend.shutdown_vm t.vm_backend) vm
+        | Vm_stop_mode.Suspend  -> t.log (Printf.sprintf "VM suspend: %s\n" vm_name);
+          or_vm_backend_error "Unable to suspend VM" (Vm_backend.suspend_vm t.vm_backend) vm
+        | Vm_stop_mode.Destroy  -> t.log (Printf.sprintf "VM destroy: %s\n" vm_name) ; 
+          or_vm_backend_error "Unable to destroy VM" (Vm_backend.destroy_vm t.vm_backend) vm
+      end
+    | Vm_state.Off
+    | Vm_state.Paused 
+    | Vm_state.Suspended
+    | Vm_state.Unknown -> Lwt.return_unit (* VM already stopped or nothing we can do... *)
 
-let start_vm t vm =
-  let state = get_vm_state t vm in
-  t.log (Printf.sprintf "Starting %s (%s)" vm.vm_name (string_of_vm_state state));
-  match state with
-  | Libvirt.Domain.InfoPaused | Libvirt.Domain.InfoShutdown
-  | Libvirt.Domain.InfoShutoff ->
-    let () = match state with
-      | Libvirt.Domain.InfoPaused ->
-        t.log " --> resuming vm...\n";
-        resume_vm t vm
-      | _ ->
-        t.log " --> creating vm...\n";
-        create_vm t vm
+  let start_vm t vm =
+    get_vm_name t vm >>= fun vm_name ->
+    get_vm_state t vm >>= fun vm_state ->
+    t.log (Printf.sprintf "Starting %s (%s)" vm_name (Vm_state.to_string vm_state));
+    let update_stats () =
+      Irmin_backend.set_start_timestamp t.storage ~vm_name (Unix.time ()) >>= fun () ->
+      Irmin_backend.inc_total_starts t.storage ~vm_name
     in
-    (* Notify Synjitsu *)
-    (match vm.mac with
-      | Some m -> 
-              (match t.synjitsu with
-              | Some s -> (
+    let notify_synjitsu () =
+      Irmin_backend.get_ip t.storage ~vm_name >>= fun r ->
+      match r with
+      | Some ip -> begin
```

Comment:
this pattern matching is a bit awful :p 

You can either do a top-level one:

```ocaml
match r, vm_mac, t.synjitsu, with
| Some ip, Some m, Some s -> [...]
| _ -> Lwt.return_unit
```
or use a function to compose them.